File: shixi/C++code/判断线段AB与线段CD是否有交点.cpp
#include <stdio.h>

struct Point{
    double x;
    double y;
};

double mult(Point a, Point b, Point c){
    return (b.y-c.y)*(a.x-c.x) - (b.x-c.x)*(a.y-c.y);
}

bool solution(Point aa, Point bb, Point cc, Point dd){
    if(mult(aa, cc, dd)*mult(bb, cc, dd)>0){
        return false;
    }
    if(mult(cc, aa, bb)*mult(dd, aa, bb)>0){
        return false;
    }
    return true;
}

int main(){
    Point aa = {0.0, 0.0};
    Point bb = {10.0, 10.0};
    Point cc = {0.0, 0.0};
    Point dd = {10.0, -10.0};
    bool ret;
    ret = solution(aa, bb, cc, dd);
    printf("%d\n", ret);
    return 0;
}

File: shixi/python_code/判断线段AB与线段CD是否有交点.py

class Solution:
    def intersection(self, start1, end1, start2, end2)-> bool:
        x1, y1, x2, y2, x3, y3, x4, y4 = *start1, *end1, *start2, *end2
        det = lambda x, y, x0, y0, x1, y1: (y1-y0)*(x-x1) - (x1-x0)*(y-y1)
        if det(x1,y1,x3,y3,x4,y4) * det(x2,y2,x3,y3,x4,y4) > 0:
            return False
        if det(x3,y3,x1,y1,x2,y2) * det(x4,y4,x1,y1,x2,y2) > 0:
            return False
        return True
        
if __name__ == "__main__":
    sol = Solution()
    ret = sol.intersection([0.0, 0.0], [10.0, 10.0], [0.0, 0.0], [10.0, -10.0])
    print(ret)
    